### DISCUSSION TITLE
conform.nvim: lsp_fallback = false not working

### DISCUSSION THREAD

**Discussion by gmnz** [LazyVim/LazyVim#2615] [Q&A]
Hi, my config file for conform.nvim looks like this:

```lua
return {
  "stevearc/conform.nvim",
  opts = {

    format = {
      lsp_fallback = false,
    },

    formatters = {
      stylua = {
        condition = function(self, _)
          return false
        end
      },
    }
  }
}
```

Why are my lua files still being autoformatted on save, please? I know I could just disable autoformatting through vim.b.autoformat, but I don't want that. I want to use stylua with a condition (of course not the one I'm using in this example), I don't want lua_ls to format my files.

**Comment by dpetka2001** [comment]
This is not configurable at the moment. Just made a [PR](https://github.com/LazyVim/LazyVim/pull/2616) to hopefully make it configurable by the user. In the meantime, you would have to disable LSP formatting individually for each server that you want like the following example
```lua
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        lua_ls = {
          settings = {
            Lua = {
              format = {
                enable = false,
              },
            },
          },
        },
      },
    },
  },
```
You would have to check the settings for other LSP servers and make corresponding changes accordingly.

**Reply by gmnz** [reply]
Oh, OK. I thought it should work from what i read here: https://github.com/LazyVim/LazyVim/issues/1790

**Reply by gmnz** [reply]
And here it also sounds to me like it should work: https://www.lazyvim.org/plugins/formatting

opts.format: extra options passed to require("conform").format(options)

**Reply by dpetka2001** [reply]
You can see [here](https://github.com/LazyVim/LazyVim/blob/91126b9896bebcea9a21bce43be4e613e7607164/lua/lazyvim/util/lsp.lua#L117-L119) that the LSP format function is setting `lsp_fallback` explicitly to `true` when `conform.nvim` is available.

**Reply by gmnz** [reply]
OK, thanks a lot for your help! :-)

**Reply by dpetka2001** [reply]
You can try out my PR locally and see it that makes any difference for you.

**Reply by dpetka2001** [reply]
You can try it by using the following spec for LazyVim
```lua
-- ~/.config/nvim/lua/plugins/core.lua
return {
  {
    "dpetka2001/LazyVim",
    branch = "fix/conform_lsp_fallback",
  },
}
```
and then check how `conform` behaves by setting `lsp_fallback` to `true` and `false`. After you create the file with the new spec for LazyVim, you should open the `Lazy` UI and press `u` on `LazyVim`, so that it gets updated to my branch. After you finish, you can discard the changes.

**Reply by gmnz** [reply]
@dpetka2001 Your PR seems to be working just like it should. Thank you! :-)

**Reply by gmnz** [reply]
I don't want to ask too much, but how difficult would it be to make this lsp_fallback setting configurable per filetype? Cause that would be really great.

**Reply by dpetka2001** [reply]
Not too familiar with conform's codebase and not too fluent with programming myself (more like a hobbyist who likes to tinker with stuff and just a user like you), so that would be a request better directed at conform's repo if that feature can be implemented. Maybe it can be done, but I'm not aware of it cuz I don't use such a case? LazyVim just uses it cuz conform is the one that provides it.

**Reply by gmnz** [reply]
Thanks, I'll ask about it in conform's repo as you suggested.

**Reply by gmnz** [reply]
I found this https://github.com/stevearc/conform.nvim/issues/129#issuecomment-1757918884 but I don't know if it's applicable for LazyVim because LazyVim doesn't expose format_on_save for modifications.... I tried something similar in "format", but that one si run only once when conform is set up.

**Reply by dpetka2001** [reply]
After some experimenting you can do the following using my branch
```lua
local filetypes = { lua = false, html = true }
local function fallback()
  local buf = vim.api.nvim_get_current_buf()
  for ft, cond in pairs(filetypes) do
    if ft == vim.bo[buf].filetype then
      return cond
    end
  end
  return true  -- default value if filetype not in table filetypes
end

return {
  {
    "stevearc/conform.nvim",
    opts = function(_, opts)
      opts.format.lsp_fallback = fallback()
      opts.formatters = {
        stylua = {
          condition = function(self, _)
            return false
          end,
        },
      }
    end,
  },
}
```
Please test it out if this configuration fits your needs.

**Reply by dpetka2001** [reply]
Or you could even use the implementation from the post you mentioned
```lua
local lsp_fallback = setmetatable({
  html = "always",
  lua = false,
}, {
  -- default true
  __index = function() return true end,
})

return {
  {
    "stevearc/conform.nvim",
    opts = function(_, opts)
      opts.format.lsp_fallback = lsp_fallback[vim.bo[vim.api.nvim_get_current_buf()].filetype]
      opts.formatters = {
        stylua = {
          condition = function(self, _)
            return false
          end,
        },
      }
    end,
  },
}
```
Whatever suits your taste.

**Reply by gmnz** [reply]
Cool, thanks! I'll test this out later.

**Reply by gmnz** [reply]
It's working, thank you again for your help.

**Reply by gmnz** [reply]
After testing it some more I have a feeling that the fallback() function is run only once when the plugin is being set up, so the value of lsp_fallback doesn't change with filetype when switching between buffers with different filetypes.

**Reply by dpetka2001** [reply]
Yes that is correct. The example from conform repo uses `format_on_save` which takes the current buffer as an argument and everything happens for the current buffer. Unfortunately, I'm unable to make it work in the PR with the current format function that LazyVim uses as my knowledge is limited. 

As a workaround, I created an autocmd that reloads `conform.nvim` each time you save (to imitate the `format_on_save` function), so that the current buffer filetype can be correctly taken into account. 
```lua
-- ~/.config/nvim/lua/config/autocmds.lua
vim.api.nvim_create_autocmd("BufWritePre", {
  group = vim.api.nvim_create_augroup("ResetConform", { clear = true }),
  callback = function()
    require("lazy.core.loader").reload("conform.nvim")
  end,
})
```
Please test out the current behavior with this and if you also happen to have a more correct approach how this could be incorporated in the PR feel free to make suggestions, I'm open to alternatives.

**Reply by gmnz** [reply]
Thanks for the suggestion, but you're right, reloading the plugin before every write is not an ideal solution. I'll try to find a better solution when I have time. For now I'm toggling the value of vim.b.autoformat before write depending on the current buffer filetype.

**Reply by caschbre** [reply]
After upgrading LazyVim and using conform, I've been unable to disable LSP formatting for Intelephense. Prior to the upgrade it was working. I'm able to alter the format settings so I know the settings are being picked up... but for some reason the format.enable is not having any effect.

```
return {
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      intelephense = {
        settings = {
          intelephense = {
            diagnostics = { enable = false },
            format = { enable = false },
          },
        },
      },
    },
  },
}
```

**Reply by gmnz** [reply]
@caschbre I'm not sure if all LSP's use the same same configuration structure (table), I think they don't. If intelephense is autoformatting your code, that means you have no formatter set up for conform.nvim, so I guess you don't want to autoformat php code at all. In that case you can use this: https://www.lazyvim.org/configuration/tips#disable-autoformat-for-some-buffers, just replace "lua" with "php".

**Reply by dpetka2001** [reply]
Not using `intelephense`, so not sure about that. What I can tell is that for `lua_ls` it works correctly doing the same thing. Are you sure Intelephense is following the schema you provided?

**Reply by caschbre** [reply]
@gmnz Yeah, I double checked the intelephense docs and it does use the same structure.

```
"intelephense.format.enable": {
    "type": "boolean",
    "default": true,
    "description": "Enables formatting.",
    "scope": "window"
},
```

Though it might be the lack of my other php formatter from working. I use phpcbf... which has been working until the change to conform. 

```
return {
  "stevearc/conform.nvim",
  opts = {
    log_level = vim.log.levels.DEBUG,
    formatters_by_ft = {
      php = { "phpcbf" },
    },
  },
}
```

Now when I run the command :ConformInfo, I get the following...

Formatters for this buffer:
LSP: intelephense
phpcbf ready (php) ~/.config/composer/vendor/bin/phpcbf

But if I run :LazyFormatInfo, it does not mention phpcbf at all. 

Is there a way to debug formatting?  I had assumed phpcbf was running after looking at ConformInfo, but that may not be the case.

**Reply by gmnz** [reply]
OK, so my guess was wrong, you want to use autoformatting. Then my advice was also wrong, don't set vim.b.autoformat to false. As I understand it, LazyVim uses LSP to autoformat on save only if your formatter isn't working or is blocked by a condition that return false at the moment. You can try to remove intelephense and autoformat on save. If it works, then conform is using phpcbf, if not, it was using intelephense because phpcbf didn't work for some reason. Or you can try using pretty-php instaed of phpcbf, I'm using it and it seems it's working fine.

**Reply by dpetka2001** [reply]
@gmnz 

Could you please try the following?
```lua
-- ~/.config/nvim/lua/util/conform.lua
local M = {}

M.fallback = setmetatable({
  html = "always",
  lua = false,
  rust = true,
  python = true,
}, {
  -- default true
  __index = function()
    return true
  end,
})

return M
```
```lua
-- ~/.config/nvim/lua/config/autocmds.lua
vim.api.nvim_create_autocmd("BufWritePre", {
  group = vim.api.nvim_create_augroup("Reset_conform_lsp_fallback", { clear = true }),
  desc = "Reset conform lsp_fallback",
  callback = function()
    local opts = require("lazyvim.util").opts("conform.nvim")
    local fallback = require("util.conform").fallback
    opts.format.lsp_fallback = fallback[vim.bo[vim.api.nvim_get_current_buf()].filetype]
  end,
})
```
and check the behavior if it is now how you want it? Create the directory and file `/lua/util/conform.lua` yourself in your personal configuration.

Edit: I'm basically changing only the `lsp_fallback` based on the filetype of each buffer before every save file instead of reloading the whole plugin. I believe this is more close to how `format_on_save` works, since it also creates an autocmd on `BufWritePre` that changes the `lsp_fallback` based on the filetype of the buffer.
